### ISSUE TITLE
Publish a lock file as part of GH releases

### ISSUE THREAD

**Issue by dominykas** [semantic-release/github#162]
Reasoning:
- lock files are useful for reproducible builds and debugging
- some people find lock files annoying, esp. when maintaining libraries

To get the best of both worlds, it would be nice to keep a lock file, but not make it a part of the repo nor a part of the npm published package.

So I'd suggest:
- when creating a GH release
- if there is a package-lock.json or an npm-shrinkwrap.json (or a yarn.lock?) - back it up
- create an npm-shrinkwrap.json (incl. dev deps) and include as an artifact
- restore the previous lock file (or delete, if none existed), so that other plugins are unaffected

I think it would be a good default, with an opt-out, but opt-in is probably also better than nothing. Not sure if there's a way to have this as a recipe (if there is reluctance to have it built in)? Maybe create it in `postpack`?

Happy to PR, if an approach can be agreed.

**Comment by pvdlg** [comment]
> To get the best of both worlds, it would be nice to keep a lock file, but not make it a part of the repo nor a part of the npm published package.

Lockfiles are never part of the package publish on npm (the shrinkwrap file is though). This why a lockfile can create discrepancy between your dev environment and the user execution environment.

> create an npm-shrinkwrap.json (incl. dev deps) and include as an artifact

What would a lockfile do as an artifact in a GitHub release?
I don't understand what you are asking here...Lockfiles are used by npm when running `npm install` from your cloned repo. Adding the file in a GitHub release will do nothing....

**Comment by dominykas** [comment]
> It also seems you are confused between lockfiles and shrinkwrap files.

When I say "lockfile" I take it to mean either - `yarn.lock`, `npm-shrinkwrap.json`, `package-lock.json`. `package-lock.json` and `npm-shrinkwrap.json` are identical in their format and contents, except one is published on npm, but since I'm proposing something different, it's irrelevant which one is used.

> I don't understand what you are asking here...Lockfiles are used by npm when running `npm install` from your cloned repo. Adding the file in a GitHub release will do nothing....

Yes, that's exactly the idea. Do not use them when installing. That's the whole point.

It is only there to keep a permanent record of versions of packages that were used to build a thing. If someone does need to use the exact set of versions - they can fetch the file and use it. They can also use it to compare the deps that were used at the build time with the deps that are available now to debug any issues, but without forcing that onto end users and therefore taking advantage of semver based updates.

In my work environment we push our applications to production using docker containers (i.e. they are immutable, final build artefacts, so the lock files don't matter at all), but sometimes we need to patch those, and for that we need reproducible builds. The only way to achieve reproducible builds is to have a lock file.

However for our daily dev purposes lockfiles are annyoing, they are hard to explain (even your assumption that you need to explain it to me kind of proves that there's a lot of people who don't understand them), they require management (even if said management is automated via greenkeeper or renovate) and attention (aka wasting time for low value chores). But I _really_ don't want to get into _this_ discussion. I'm well aware of pros/cons, and it's clearly not a green/red choice.

At work, we have currently solved this by saving the `npm-shrinkwrap.json` in a dangling tag, and if we need to publish a patch on an existing release - we start off that tag. But that's for work, and it's also a workaround. GH releases is a good place to store build artefacts, and a snapshot of dependencies is exactly that.

For open source node modules, there are also [discussions about the value of reproducible builds](https://github.com/nodejs/package-maintenance/issues/77), which is where the lock file question comes in. There's people who are averse to lockfiles in npm or git, and they have the right to that opinion, because it's completely valid, so having a lockfile published on GH as part of GH release, seems like a winner.

> That's why it's recommend to not use lockfile for libraries published on npm, but use them for application that are run via `npm install && npm run start`.

"it's recommended" is a rather broad statement. Different people recommend different things. There are plenty of developers, in open source and otherwise, who recommend to never use the `package-lock.json` at all. The whole "shrinkwrap for apps" is also debatable - it's a pointless overhead if your app is already packaged up (zeit/pkg, electron, docker, etc etc). And there are open source developers (e.g. hapi), who publish their library with a shrinkwrap, so that they can guarantee a full set of dependencies, because that's their promise of quality.

There's different needs and different contexts, and I think it's unfair to say that the lock file question is a settled recommendation.

**Comment by dominykas** [comment]
Or, let me rephrase the original issue too, hopefully this will explain things better.

Saving a shrinkwrap/lock file is a good idea for libs _and_ apps, if only for future reference. With the proposed feature, the maintainers of the lib/app could:

1. Use `package-lock.json`, if they like it, but would also get `npm-shrinkwrap.json` in GH releases (yes, that's duplication)
2. Use `npm-shrinkwrap.json`, if they l like it, but would also get `npm-shrinkwrap.json` in GH releases (yes, that's duplication)
3. Not use any of the two, but would get `npm-shrinkwrap.json` in GH releases, which might help _other_ people, as well as the maintainer, in case they do need it* 

Having `npm-shrinkwrap.json` consistently in releases for all three cases results in duplication, but provides a benefit of always knowing where to find it.

* Fun fact: A couple of years ago, I had a couple of chances to manually rebuild an `npm-shrinkwrap.json` from an `npm ls` seen in the logs. It was not exactly something I'd define as "fun".

**Comment by pvdlg** [comment]
Thanks for the thorough explanations @dominykas!

I think your use case is not common enough to implement it as default behaviour in semantic-release.
You can however achieve that with the current plugins:
- Use `@semantic-release/exec` to generate your `npm-shrinkwrap.json`  and move it in a temp folder
- Use the the `assets` option of `@semantic-release/github` to add it to the GH release

**Comment by dominykas** [comment]
Thanks, I understand. I was thinking of a wider application for reproducible build tooling, but I guess we'll have to wait and see.

I'm also reluctant to tinker with `plugins` defaults (not sure why, just doesn't feel comfortable, maybe cause so much of default functionality depends on that), but I haven't looked at shared configs, so maybe there's scope for easy setup that way.
